File: web/src/services/documents.ts
export type DocumentStatus = "uploading" | "parsing" | "indexing" | "ready" | "failed";

export type RagDocument = {
  id: string;
  name: string;
  file_hash: string;
  mime_type: string;
  size: number;
  status: DocumentStatus;
  error_message: string | null;
  version: number;
  permission_tags: string[];
  created_by: string | null;
  created_at: string;
  updated_at: string;
};

export type DocumentListResponse = {
  items: RagDocument[];
  total: number;
  page: number;
  page_size: number;
};

const apiUrl = import.meta.env.VITE_SSW_API_URL || "http://127.0.0.1:8000";

async function responseError(response: Response, fallback: string): Promise<Error> {
  try {
    const payload = await response.json() as { detail?: unknown; message?: unknown };
    if (typeof payload.detail === "string") {
      return new Error(payload.detail);
    }
    if (typeof payload.message === "string") {
      return new Error(payload.message);
    }
  } catch {
    // 非 JSON 错误响应使用统一提示。
  }
  return new Error(`${fallback}：${response.status}`);
}

export async function listDocuments(
  page = 1,
  pageSize = 20,
  status?: DocumentStatus,
): Promise<DocumentListResponse> {
  const params = new URLSearchParams({
    page: String(page),
    page_size: String(pageSize),
  });
  if (status) {
    params.set("status", status);
  }

  const response = await fetch(`${apiUrl}/documents?${params.toString()}`);
  if (!response.ok) {
    throw await responseError(response, "加载文档失败");
  }
  return await response.json() as DocumentListResponse;
}

export async function uploadDocument(file: File): Promise<RagDocument> {
  const formData = new FormData();
  formData.append("file", file);

  const response = await fetch(`${apiUrl}/documents`, {
    method: "POST",
    body: formData,
  });
  if (!response.ok) {
    throw await responseError(response, "上传文档失败");
  }
  return await response.json() as RagDocument;
}

export async function retryDocument(documentId: string): Promise<RagDocument> {
  const response = await fetch(`${apiUrl}/documents/${documentId}/retry`, {
    method: "POST",
  });
  if (!response.ok) {
    throw await responseError(response, "重试文档失败");
  }
  return await response.json() as RagDocument;
}

export async function deleteDocument(documentId: string): Promise<void> {
  const response = await fetch(`${apiUrl}/documents/${documentId}`, {
    method: "DELETE",
  });
  if (!response.ok) {
    throw await responseError(response, "删除文档失败");
  }
}

File: src/ssw/api/documents.py
"""文档管理 API：上传、列表、详情、删除、重试、下载/预览、chunk 浏览、改权限标签、重新索引。

第 11 章接入认证与权限：
- 读接口（list / get / chunks / download）→ CurrentUser，按 permission_tags 过滤
- 写接口（upload / delete / retry / update tags / reindex）→ CurrentAdmin
"""


from urllib.parse import quote
from uuid import UUID

from fastapi import APIRouter, File, Query, Response, UploadFile
from starlette.requests import Request

from ssw.db.models import Document, DocumentStatus
from ssw.dependency import DocumentServiceDep
from ssw.schemas.documents import DocumentRead, DocumentStatusValue, DocumentListResponse, \
    DocumentChunkListResponse, DocumentChunkRead, DocumentChunkStats, DocumentChunkDetail

documents_router = APIRouter(prefix="/documents", tags=["documents"])

# DOCX 即便 ?download=0 也强制 attachment：浏览器无法内联渲染 DOCX
_DOCX_MIME = "application/vnd.openxmlformats-officedocument.wordprocessingml.document"



@documents_router.post("", response_model=DocumentRead, status_code=201, operation_id="uploadDocument")
async def upload_document(
    service: DocumentServiceDep,
    http_request: Request,
    file: UploadFile = File(..., description="待上传文档（PDF / DOCX / Markdown / HTML）")
) -> DocumentRead:
    """上传文档：写入 COS、落库后立即返回，解析与向量化由 Celery worker 异步执行。"""

    document = await service.upload(
        file,
        created_by=http_request.state.user_id
    )
    return _to_document_read(document)


@documents_router.get("", response_model=DocumentListResponse, operation_id="listDocuments")
async def list_documents(
    service: DocumentServiceDep,
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    status: DocumentStatusValue | None = Query(None, description="按文档状态筛选"),
) -> DocumentListResponse:
    items, total = await service.list_documents(
        page,
        page_size,
        status=DocumentStatus(status) if status else None,
    )
    return DocumentListResponse(
        items=[_to_document_read(d) for d in items],
        total=total,
        page=page,
        page_size=page_size,
    )


@documents_router.get("/{document_id}", response_model=DocumentRead, operation_id="getDocument")
async def get_document(
    document_id: UUID,
    service: DocumentServiceDep,
) -> DocumentRead:
    document = await service.get(document_id)
    return _to_document_read(document)


@documents_router.delete("/{document_id}", status_code=204, operation_id="deleteDocument")
async def delete_document(
    document_id: UUID,
    service: DocumentServiceDep,
) -> Response:

    await service.delete(document_id)
    return Response(status_code=204)


@documents_router.post(
    "/{document_id}/retry",
    response_model=DocumentRead,
    operation_id="retryDocument",
)
async def retry_document(
    document_id: UUID,
    service: DocumentServiceDep,
) -> DocumentRead:
    document = await service.retry(document_id)
    return _to_document_read(document)


@documents_router.post(
    "/{document_id}/reindex",
    response_model=DocumentRead,
    operation_id="reindexDocument",
)
async def reindex_document(
    document_id: UUID,
    service: DocumentServiceDep,
    file: UploadFile = File(
        ..., description="新版本文件（MIME 必须与原文档一致）"
    ),
) -> DocumentRead:
    """上传新版本文件，触发按 chunk_hash 对齐的增量重建。"""
    document = await service.reindex(document_id, file)
    return _to_document_read(document)





@documents_router.get("/{document_id}/file", operation_id="downloadDocument")
async def download_document(
    document_id: UUID,
    service: DocumentServiceDep,
    download: int = Query(0, ge=0, le=1, description="1=强制下载, 0=尝试内联预览"),
) -> Response:
    """返回文档原始字节。

    支持两种认证方式（优先 header）：
    - Authorization header（常规 fetch 请求）
    - ?token= query 参数（iframe / 浏览器直接打开时无法带 header）
    """


    document = await service.get(document_id)
    content = await service.file_service.download(document.oss_object_key)

    force_attachment = download == 1 or document.mime_type == _DOCX_MIME
    disposition = "attachment" if force_attachment else "inline"
    # RFC 5987 编码非 ASCII 文件名，避免中文文件名报错
    filename_quoted = quote(document.name, safe="")

    return Response(
        content=content,
        media_type=document.mime_type,
        headers={
            "Content-Disposition": (
                f"{disposition}; filename*=UTF-8''{filename_quoted}"
            ),
        },
    )


@documents_router.get(
    "/{document_id}/chunks",
    response_model=DocumentChunkListResponse,
    operation_id="listDocumentChunks",
)
async def list_document_chunks(
    document_id: UUID,
    service: DocumentServiceDep,
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
) -> DocumentChunkListResponse:
    items, total, stats = await service.list_chunks(
        document_id, page, page_size
    )
    return DocumentChunkListResponse(
        items=[DocumentChunkRead.from_orm_chunk(c) for c in items],
        total=total,
        page=page,
        page_size=page_size,
        stats=DocumentChunkStats(
            total=stats.total,
            avg_length=stats.avg_length,
            min_length=stats.min_length,
            max_length=stats.max_length,
        )
        if stats is not None
        else None,
    )


@documents_router.get(
    "/{document_id}/chunks/{chunk_id}",
    response_model=DocumentChunkDetail,
    operation_id="getDocumentChunk",
)
async def get_document_chunk(
    document_id: UUID,
    chunk_id: UUID,
    service: DocumentServiceDep
) -> DocumentChunkDetail:
    chunk = await service.get_chunk(
        document_id, chunk_id
    )
    return DocumentChunkDetail.from_orm_chunk(chunk)


def _to_document_read(document: Document) -> DocumentRead:
    return DocumentRead.model_validate(document)
